### ISSUE TITLE
Cookie/session authentication without cookie header

### ISSUE THREAD

**Issue by Prinzhorn** [cloudant/nodejs-cloudant#112]
I have a use case which I think is not covered by Cloudant, but should be very common when using PouchDB or any client side CouchDB client.

I use a one-database-per-project approach. Users create projects and edit them in a browser editor. The editor syncs the changes with the remote Cloudant database. I love this approach as I can easily copy/clone projects using replication and give multiple users access to a project using _security.

But the question is, how do I authenticate the browser against Cloudant? Two options
1. Send plaintext user/password to the browser so it can send it to Cloudant via basic auth -> no way I'm going to do this.
2. Create a Cloudant session on the server and send the Cookie to the browser -> doesn't work because I can't set a third party cookie for the CORS request.

In order for number 2 to work, I need to be able to send the `AuthSession` in a custom header which I can control in the xhr instance.

Now that I think about it, this wouldn't work either. Because Cloudant doesn't use _users (is it fully supported if I enable it?). What I want to avoid is that my server side application has access to plaintext passwords. The only exception is during signup (in which case CouchDB would do the password hashing for me, Cloudant doesn't) and login (which would be the only place I'm able to request a session from Cloudant and my application would only pass the credentials over without doing anything with them).

So, how do I solve this? Even if I roll my own user management (which I don't want) and generate an API key per user, I still have two problems: I'd need to store the API key in plaintext/encrypted, because I need it in plaintext to get a session from Cloudant and additionally I sill can't send it as a CORS cookie.

The only option I currently see is rolling my own sessions by generating API keys and expiring them within my application logic (which would mean constantly updating the _security of the databases).

Edit: of course there's at least one other option, which would be to proxy all db traffic through my backend (which has its own session management), but that's pointless as I want the DB to scale independent from my application. The PouchDB communication with Cloudant is like 99% of the traffic, the application server just glues everything together.

**Comment by glynnbird** [comment]
This is a great question. You could use the PouchDB auth plugin https://github.com/nolanlawson/pouchdb-authentication

This assumes you're speaking to a CouchDB server so you'll have to make your Cloudant account behave like CouchDB, as described in the last paragraph of this post: https://cloudant.com/for-developers/faq/auth/

**Comment by Prinzhorn** [comment]
Thanks for the quick reply!

I'm aware of this plugin, but I still have a backend which does things like signup. It's impossible to call `db.login(username, password [, options] [ callback])` on the client without either sending the plaintext credentials to the browser first (insecure) or asking the user for credentials a second time (bad ux). I _could_ call `db.login` on my login form before submitting it, but that's a hack and nothing more. It would also mean I have two separate sessions. And I can't do things like sending the user a link via email which then logs the user in.

So I basically want the browser to communicate with Cloudant directly with the exception of authentication, which should exclusively happen on the server.

**Comment by Prinzhorn** [comment]
I'm currently evaluating my options as I need a solution early next year. I want to stick with CouchDB in general for the reasons mentioned above (PouchDB sync) even though I looked at other databases (but I don't want to roll conflict managament/revisions on my own). As far as I understand even running my own CouchDB won't help, because I can't set cookies from `www.example.com` to `couch.example.com` (without polluting the whole namespace and setting the cookie to `.example.com`).

I came to the conclusion that proxying the db requests through my backend is the **only** option available. In general that's not a bad thing, as it's the way it would work with every other database (e.g. Mongo or Rethink) but on the other hand that's the point of CouchDB to have it exposed directly via HTTP.

Maybe someone else has an idea? I was also searching for production sites using Pouch + Couch and couldn't find any (except for Couch apps, which don't have a backend as I do). I was curious how they solved it.

**Comment by dlucian** [comment]
@Prinzhorn, I'm curious on what solution did you come up with? I'm facing quite the same challenge, and I'm probably going for a backend that handles Auth.

**Comment by glynnbird** [comment]
I feel I should mention this project in this context too:

https://github.com/cloudant-labs/envoy

tl;dr your client side app is still "one db per user" but at the server side, all the content is in one database for ease of querying every user's data for analytics. It also handles auth too. Might be worth a look.

**Comment by Prinzhorn** [comment]
@dlucian no solution. I went with RethinkDB and a proper application layer.
